src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,257 @@
+        {
+            FinishEvent();
+        }
+    }
+
+    private bool IsSurfacePatch(Patch patch)
+    {
+        return patch.Depth[0] == 0 && patch.BiomeType != BiomeType.Cave;
+    }
+
+    private void MarkPatches(double totalTimePassed)
+    {
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (IsSurfacePatch(patch))
+            {
+                patch.AddPatchEventRecord(WorldEffectVisuals.GlobalGlaciation, totalTimePassed);
+            }
+        }
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (!AreConditionsMet())
+            return;
+
+        eventDuration = random.Next(Constants.GLOBAL_GLACIATION_MIN_DURATION, Constants.GLOBAL_GLACIATION_MAX_DURATION);
+        generationsLeft = eventDuration;
+
+        foreach (var (index, patch) in targetWorld.Map.Patches)
+        {
+            if (IsSurfacePatch(patch))
+            {
+                ChangePatchProperties(index, patch, totalTimePassed);
+            }
+        }
+
+        LogBeginningOfGlaciation();
+    }
+
+    private bool AreConditionsMet()
+    {
+        var numberOfSurfacePatches = 0;
+        var patchesExceedingOxygenLevel = 0;
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (!IsSurfacePatch(patch))
+                continue;
+
+            patch.Biome.TryGetCompound(Compound.Oxygen, CompoundAmountType.Biome, out var oxygenLevel);
+            if (oxygenLevel.Ambient >= Constants.GLOBAL_GLACIATION_OXYGEN_THRESHOLD)
+                patchesExceedingOxygenLevel += 1;
+
+            numberOfSurfacePatches += 1;
+        }
+
+        // Just prevent dividing by zero, but that shouldn't be possible anyway
+        numberOfSurfacePatches = numberOfSurfacePatches == 0 ? 1 : numberOfSurfacePatches;
+
+        return (float)patchesExceedingOxygenLevel / numberOfSurfacePatches >=
+            Constants.GLOBAL_GLACIATION_PATCHES_THRESHOLD
+            && random.NextFloat() <= Constants.GLOBAL_GLACIATION_CHANCE;
+    }
+
+    private void ChangePatchProperties(int index, Patch patch, double totalTimePassed)
+    {
+        modifiedPatchesIds.Add(index);
+        AdjustBackground(patch);
+        AdjustEnvironment(patch);
+        AddIceChunks(patch);
+        LogEvent(patch, totalTimePassed);
+    }
+
+    private void AdjustBackground(Patch patch)
+    {
+        patch.CurrentSnapshot.Background = Background;
+    }
+
+    private void AdjustEnvironment(Patch patch)
+    {
+        bool hasTemperature =
+            patch.Biome.ChangeableCompounds.TryGetValue(Compound.Temperature, out var currentTemperature);
+        bool hasSunlight = patch.Biome.ChangeableCompounds.TryGetValue(Compound.Sunlight, out var currentSunlight);
+
+        if (!hasTemperature)
+        {
+            GD.PrintErr("Patch has no temperature");
+            return;
+        }
+
+        if (!hasSunlight)
+        {
+            GD.PrintErr("Patch has no sunlight");
```

Comment:
Is this actually an error? Patches do exist that are allowed to have no sunlight.

Reply:
Yes, but all surface patches must have light. I will update the text to make it more obvious

Reply:
Yeah, I did see that the code calling this makes sure these are surface patches. Though it isn't totally optimal that an error would only be triggered once the event tries to start, if there is a new bug introduced in the map generator that accidentally leaves out sunlight.

Reply:
So do you want the error print removed and to just leave the `return` statement?

Reply:
I mean, don't event depth patcch have light level which equals 0? I think that this error print should be there

Reply:
I guess it is better to leave the error in but I think it should be worded more like "Glaciation event encountered patch with unexpectedly no sunlight" like that to emphasise it's a guard clause kind of thing and based on assumptions in the glaciation event code.